### PR TITLE
enable hypothesis testing to output a random number of tensors

### PR DESCRIPTION
**Pull request by hyuen**
Summary: we needed a random number of tensors with the same shape to write some tests, enhancing hypothesis to allow for this

Differential Revision: D16823566

### CODE REVIEW

File: caffe2/python/hypothesis_test_util.py [hunk 1]
```diff
@@ -270,6 +270,15 @@ def tensors(n, min_dim=1, max_dim=4, dtype=np.float32, elements=None, **kwargs):
             max_size=n))
 
 
+def ntensors(min_n, max_n, min_dim=1, max_dim=4, dtype=np.float32, elements=None, **kwargs):
+    dims_ = st.lists(dims(**kwargs), min_size=min_dim, max_size=max_dim)
```
Comment (yinghai) [inline caffe2/python/hypothesis_test_util.py]: Please add docstring on what this function is about.

Comment (hyuen) [reply caffe2/python/hypothesis_test_util.py]: updated

**Comment by yinghai** [comment]
Can you check CI failures?

**Review by soumith — DISMISSED** [review]

**Comment by salexspb** [comment]
could you add example usage? I see only new code and it is not used anywhere. Hard to see if it is working or not. If you are doing a stack, could you use gh stack or manually show usage example in the diff description?

**Comment by hyuen** [comment]
@yinghai 
windows failure: failed to clean the workspace (environmental)

the other three tests: _dr.ci, caffe2_onnx, pytorch_linux, they all fail because:
data_parallel_model_test.py::RecurrentNetworkParallelTest::test_equiv_recurrent failed

I don't think this change is dependent on it

@salexspb for use cases this is to test a sum of random numbers of tensors, the existing infrastructure only allows us to do fixed numbers of tensors

File: caffe2/python/hypothesis_test_util.py [hunk 2]
```diff
@@ -270,6 +270,19 @@ def tensors(n, min_dim=1, max_dim=4, dtype=np.float32, elements=None, **kwargs):
             max_size=n))
 
 
+def ntensors(min_n, max_n, min_dim=1, max_dim=4, dtype=np.float32, elements=None, **kwargs):
+    """
+    Create between min_n to max_n tensors, all of them with the same dimensions
+    which will be between [min_dim, max_dim] of type dtype.
```
[the remark this answers is not in the document]
Comment (hyuen) [reply caffe2/python/hypothesis_test_util.py]: added

File: caffe2/python/hypothesis_test_util.py [hunk 3]
```diff
@@ -270,6 +270,19 @@ def tensors(n, min_dim=1, max_dim=4, dtype=np.float32, elements=None, **kwargs):
...
+    """
+    dims_ = st.lists(dims(**kwargs), min_size=min_dim, max_size=max_dim)
```
[the remark this answers is not in the document]
Comment (hyuen) [reply caffe2/python/hypothesis_test_util.py]: they are unbounded, but typically we test with dimensions 1 to 4, that's how all caffe2 code is tested

**Pull request closed without merging**